### DISCUSSION TITLE
poetry python version is not detected on dependency extraction

### DISCUSSION THREAD

**Discussion by kota65535** [renovatebot/renovate#19144] [Request Help]
In pyproject.toml, I have following poetry dependencies section.

```
[tool.poetry.dependencies]
python = "3.10.4"
fastapi = "^0.87.0"

...
```

But Renovate does not detect the python version as dependency. Is this expected behavior?

**Comment by rarkins** [comment]
What does the `packageFiles with updates` log message say about this package file?

**Comment by kota65535** [comment]
Nothing about the python version, like following:

```
{
  "name": "renovate",
  "hostname": "SR-Mac-2006-01.local",
  "pid": 89023,
  "level": 20,
  "logContext": "vhyn0yqxEY3ItpS2F0iuA",
  "repository": "sre-aip/sample-ml",
  "baseBranch": "main",
  "config": {
    "poetry": [
      {
        "packageFile": "sample-ml-core/sample-ml-core-app/pyproject.toml",
        "deps": [
          {
            "depName": "fastapi",
            "depType": "dependencies",
            "currentValue": "0.87.0",
            "managerData": {
              "nestedVersion": false
            },
            "datasource": "pypi",
            "lockedVersion": "0.87.0",
            "versioning": "pep440",
            "depIndex": 0,
            "updates": [
              {
                "bucket": "non-major",
                "newVersion": "0.88.0",
                "newValue": "0.88.0",
                "releaseTimestamp": "2022-11-27T14:54:57.000Z",
                "newMajor": 0,
                "newMinor": 88,
                "updateType": "minor",
                "branchName": "renovate/all-minor-patch"
              }
            ],
            "warnings": [],
            "sourceUrl": "https://github.com/tiangolo/fastapi",
            "currentVersion": "0.87.0",
            "isSingleVersion": true,
            "fixedVersion": "0.87.0"
          },
          
    ...
```

**Reply by rarkins** [reply]
Yes, it appears to be intentional: https://github.com/renovatebot/renovate/blob/ebf7d638d6d731a98479f9cb6c60120c2befcb0b/lib/modules/manager/poetry/extract.ts#L53-L55

But it's detected as constraints: https://github.com/renovatebot/renovate/blob/ebf7d638d6d731a98479f9cb6c60120c2befcb0b/lib/modules/manager/poetry/extract.ts#L170-L173

I don't see a reason why we couldn't treat `python` as a dependency too, but it will need a code change. It was done this way in https://github.com/renovatebot/renovate/pull/6490 because it was at that time mistakenly trying to fetch it from pypi

**Comment by kota65535** [comment]
Thanks, that makes sense.
Poetry's python version in dependencies section means compatible python version, so Renovate treats it as a constraint.
I'm going to use pyenv .python-version file to update python version poetry uses.

**Comment by kota65535** [comment]
Hi @rarkins,
I thought about this question for a while, but I still don't get the reason why Renovate doesn't update the poetry's python version constraint.
As far as I know, Renovate allows to update terraform version constraint declared as follows:

```
terraform {
  required_version = "~> 1.3.0"
}
```

Is there any reason why we treat the version constraints of terraform and python differently?

**Reply by rarkins** [reply]
I answered above:

> I don't see a reason why we couldn't treat python as a dependency too

**Reply by kota65535** [reply]
So could this be a feature request?

**Reply by rarkins** [reply]
Or PR
